File: CLRS/counting_sort.py
def counting_sort(array_unsorted, max_value):
	array_temp = [0 for i in range(max_value + 1)]
	array_return = [float("-inf") for i in range(len(array_unsorted))]
	for i in range(len(array_unsorted)):
		array_temp[array_unsorted[i]] += 1
	for i in range(1, max_value + 1):
		array_temp[i] += array_temp[i - 1]
	"""
	print("array_unsorted",array_unsorted)
	print("array_temp",array_temp)
	print("array_return",array_return)
	"""
	for i in range(len(array_unsorted) - 1, -1, -1):
		array_return[array_temp[array_unsorted[i]] - 1] = array_unsorted[i]
		array_temp[array_unsorted[i]] -= 1
		#print (i)
		#print (array_return)
	return array_return
	
if __name__ == '__main__':
	A = [2,5,3,0,2,3,0,3]
	print(counting_sort(A, 5))

File: CLRS/counting_sort.c
#include <stdio.h>
#include <stdlib.h>
void counting_sort(int* array_unsorted, int* return_array, int max_value, int lenth_array)
{
	int* array_temp = (int *)malloc(sizeof(int) * (max_value + 1)),
		i;
	for (i = 0; i <= max_value; i++)
		array_temp[i] = 0;
	for (i = 0; i < lenth_array; i++)
		array_temp[array_unsorted[i]] = array_temp[array_unsorted[i]] + 1;
	for (i = 1; i <= max_value; i++)
		array_temp[i] += array_temp[i - 1];
	for (i = 0; i < lenth_array; i++)
	{
		return_array[array_temp[array_unsorted[i]] - 1] = array_unsorted[i];
		array_temp[array_unsorted[i]]--;
	}

	free(array_temp);
}
int mian(void)
{
	int A[8] = { 2,5,3,0,2,3,0,3 };
	const int SIZE = sizeof(A) / 4;
	int B[SIZE] = { 0 };
	counting_sort(A, B, 5, SIZE);
	for (int i = 0; i < SIZE; i++)
		printf("%d  ", A[i]);
	printf("\n");
	return 0;
}